lib/evaluation/evaluationClientWrapper.ts

```diff
@@ -0,0 +1,211 @@
+          "issuer": "",
+          "type": "VerifiableCredential"
+        }],
+      }], warnings: []
+    }
+     */
+    return null;
+  }
+  public evaluate(pd: PresentationDefinition, vp: unknown): EvaluationResults {
+    this._client.evaluate(pd, vp);
+    const result: any = {};
+    result.warnings = this._client.results
+      .filter((result) => result.status === Status.WARN)
+      .map((x) => {
+        return {
+          name: x.evaluator,
+          message: `${x.message}: ${x.input_descriptor_path}: ${x.verifiable_credential_path}`,
+        };
+      });
+    result.errors = this._client.results
+      .filter((result) => result.status === Status.ERROR)
+      .map((x) => {
+        return {
+          name: x.evaluator,
+          message: `${x.message}: ${x.input_descriptor_path}: ${x.verifiable_credential_path}`,
+        };
+      });
+    if (this._client.verifiablePresentation['presentationSubmission']['descriptor_map'].length) {
+      result.value = this._client.verifiablePresentation['presentationSubmission'];
+    }
+    return result;
+  }
+
+  public submissionFrom(pd: PresentationDefinition, vcs: unknown[]): PresentationSubmission {
+    if (!this._client.results) {
+      throw Error('You need to call evaluate() before submissionFrom()');
+    }
+    if (pd.submission_requirements) {
+      const marked: HandlerCheckResult[] = this._client.results.filter(
+        (result) =>
+          result.evaluator === 'MarkForSubmissionEvaluation' && result.payload.group && result.status !== Status.ERROR
+      );
+      for (const submissionRequirement of pd.submission_requirements) {
+        if (submissionRequirement.rule === Rules.All) {
+          this.handleAll(submissionRequirement, marked);
+        } else if (submissionRequirement.rule === Rules.Pick) {
+          this.handlePick(submissionRequirement, marked);
+        } else {
+          throw Error('Unsupported rule');
+        }
+      }
+    }
+    return this.remapVcs(vcs);
+  }
+
+  private handlePick(submissionRequirement: SubmissionRequirement, marked: HandlerCheckResult[]) {
+    if (submissionRequirement.from) {
+      const count = this.countMatchingInputDescriptors(submissionRequirement, marked);
+      this.handleCount(submissionRequirement, count);
+    } else if (submissionRequirement.from_nested) {
+      const counter: number = this.handleNested(submissionRequirement, marked);
+      this.handleCount(submissionRequirement, counter);
+    }
+  }
+
+  private handleAll(submissionRequirement: SubmissionRequirement, marked: HandlerCheckResult[]) {
+    if (submissionRequirement.from) {
+      if (this.countMatchingInputDescriptors(submissionRequirement, marked) !== marked.length) {
+        throw Error(`Not all input descriptors are members of group ${submissionRequirement.from}`);
+      }
+    } else if (submissionRequirement.from_nested) {
+      if (this.handleNested(submissionRequirement, marked) !== submissionRequirement.from_nested.length) {
```

Comment:
This is incorrect. We needed to compare with the count.
And we needed to call recursively. Because the submission requirments can be deeper than 2levels as well

Reply:
This is significant enough that we should do it carefully and exclude from Alpha

Reply:
Unless you decide to leave out the nested stuff and fix that for the beta, I think not having VPs as a result would be bad for the alpha tbh. It is too vital for the actual purpose of this libarary

Reply:
Fixed and refactored!